osu.Game/Rulesets/Scoring/HitResult.cs

```diff
@@ -291,6 +328,105 @@ public static bool IsValidHitResult(this HitResult result, HitResult minResult,
         /// <param name="result">The <see cref="HitResult"/> to get the index of.</param>
         /// <returns>The index of <paramref name="result"/>.</returns>
         public static int GetIndexForOrderedDisplay(this HitResult result) => order.IndexOf(result);
+
+        public static void ValidateHitResultPair(HitResult maxResult, HitResult minResult)
```

Comment:
I wonder if we can agree on this, and of the override-ability of `MinResult` that prompted this. I'm accepting of alternatives...

Reply:
I'm not really sure there's a feasible alternative that wouldn't require disproportionate effort...

Like, theoretically, you _could_ leverage the type system for this a _little_ bit. Separate `HitResult`s that are hits from misses. But that only saves you the first two batches of checks. The pairing of "is this a valid max result for this min result" is not really directly encodeable in the C# type system without overcomplications.

You'd need a proper typescript/haskell-like union type system to encode this nicely. Unfortunately we don't get one to work with. At least one that isn't handrolled and completely clunky to work with.

*That said*, the method itself is an absolute behemoth and can be slimmed down. The best I can do is something like:

```csharp
        public static void ValidateHitResultPair(HitResult maxResult, HitResult minResult)
        {
            if (maxResult == HitResult.None || !IsHit(maxResult))
                throw new ArgumentOutOfRangeException(nameof(maxResult), $"{maxResult} is not a valid maximum judgement result.");

            if (minResult == HitResult.None || IsHit(maxResult))
                throw new ArgumentOutOfRangeException(nameof(minResult), $"{minResult} is not a valid minimum judgement result.");

            if (maxResult.IsBonus() && minResult != HitResult.IgnoreMiss)
                throw new ArgumentOutOfRangeException(nameof(minResult), $"{HitResult.IgnoreMiss} is the only valid minimum result for a {maxResult} judgement.");

            if (maxResult == HitResult.LargeTickHit && minResult is not (HitResult.LargeTickMiss or HitResult.IgnoreMiss or HitResult.ComboBreak))
                throw new ArgumentOutOfRangeException(nameof(minResult), $"{minResult} is not a valid minimum result for a {maxResult} judgement.");

            if (maxResult == HitResult.SmallTickHit && minResult is not (HitResult.SmallTickMiss or HitResult.IgnoreMiss or HitResult.ComboBreak))
                throw new ArgumentOutOfRangeException(nameof(minResult), $"{minResult} is not a valid minimum result for a {maxResult} judgement.");

            if (maxResult.IsBasic() && minResult is not (HitResult.Miss or HitResult.IgnoreMiss or HitResult.ComboBreak))
                throw new ArgumentOutOfRangeException(nameof(minResult), $"{minResult} is not a valid minimum result for a {maxResult} judgement.");
        }
```

And yes this _slightly_ bends our ban on `or` pattern expressions but I think this is one of the rarer cases where this actually vastly benefits readability so I'd be personally fine with it.

Reply:
Hmm, the part I’m more curious about is we used to have judgements be fully customisable at one point, with no validation whatsoever. This is putting a foot back into that pond with MinResult once again being customisable.

Whether everyone’s okay with that…

An alternative I have is to separate it into a property like `OnlyComboBreakOnMiss` and keep MinResult sealed/handle everything internally.

I’m not sure if there’s any other alternatives.

Reply:
To illustrate my point here, I've added the commit https://github.com/ppy/osu/pull/25062/commits/04ce223e66d0d169500e35d636149796461f76e1, which I believe is more correct than the rules I had here. It sounds weird to have a pairing of `<Perfect, IgnoreMiss>` - I don't know when this sort of pairing would ever come about.

So, the only thing really added in this PR is the pairing `<IgnoreHit, (IgnoreMiss | ComboBreak)>`

Which, could be handled by a third alternative - making `HitResult` a flags enum and implicitly attaching ` | ComboBreak` to `IsBasic()`/`LargeTickHit` results and allowing it to be attached for `IgnoreHit`. For example...

```
class MyCustomJudgement : Judgement
{
    public override HitResult MaxResult => IgnoreHit | ComboBreakOnMiss;
}
```
Again, with the theme of keeping `MinResult` sealed.

Reply:
> the part I’m more curious about is we used to have judgements be fully customisable at one point, with no validation whatsoever. This is putting a foot back into that pond with MinResult once again being customisable.
> 
> Whether everyone’s okay with that…

I see no issue with that.

> An alternative I have is to separate it into a property like `OnlyComboBreakOnMiss` and keep MinResult sealed/handle everything internally. It would remove the need for this method altogether (could be a test, but not a live assertion).

Where would this property live? In `Judgement`? I thought that was infeasible because there are server-side components that depend on being able to calculate the max combo of a score from just the hit results alone?

I also generally don't like that flag, it feels much too specific and targeted to this weird usecase.

> making `HitResult` a flags enum and implicitly attaching ` | ComboBreak` to `IsBasic()`/`LargeTickHit` results and allowing it to be attached for `IgnoreHit`. For example...
> 
> ```
> class MyCustomJudgement : Judgement
> {
>     public override HitResult MaxResult => IgnoreHit | ComboBreakOnMiss;
> }
> ```

I _really_ don't like that one. The fact that the max result now influences min result in this roundabout way feels really bad.

Put it this way: I much prefer explicitly spelling out min/max result with appropriate validation to introducing such implicit rules.